torch/csrc/distributed/rpc/py_rref.cpp

```diff
@@ -84,6 +84,20 @@ TypePtr tryInferTypeWithTypeHint(
         "type_hint should only be specified when the RRef being created contains a ScriptModule.");
   }
 
+  // Check if value is an instance of a ScriptClass. If not, skip type inference
+  // because it will try to script the class that value is in instance of, and
+  // this should be avoided.
+  py::bool_ can_compile = py::module::import("torch._jit_internal").attr("can_compile_class")(value.get_type());
```

Comment:
These changes are necessary because `tryToInferType` now automatically scripts classes that can be scripted, which can result in scripting classes that the user did not intend to be scripted. In a JIT context, this makes more sense because you'd probably want your module attributes to all be scripted. However, at present, you can construct an `RRef` with a regular Python class instance too, and because this code uses `jit::tryToInferType`, due to the other changes in this PR (automatic scripting of class types), it is no longer possible to do this.

This small patch checks if the type of the object you are creating an `RRef` of could be a compiled as a class type and if it can, it checks if it already has been compiled or not. If it has, that means the class has been decorated with `@torch.jit.script` and the user intended it to be a TorchScript class. If not, it means that the user did not intend to script it, so we short-circuit and return `PyObjectType::get()`.

Reply:
curious if we should just enable recursive scripting to automatically compile the class? I think `tryToInferType` shouldn't automatically compile the class if user don't provide the `torch.jit.script` annotation, which make more sense to me that only explicit annotation marking the user intention to compile the class and use the object in their code. Do we have a use case why `tryToInferType` has to compile the class?

Reply:
The motivation for the PR is for automatically scripting module attributes that are classes. We do recursive scripting automatically for submodules, so this is an extension of that. I think that makes sense, and I think the problem is that this `RRef` code is using a type inference function that is meant for JIT purposes.

Reply:
Yeah I agree that we would like to apply this to module attributes that are classes. But I thought we already automatically compile it when the class attribute get annotated ([here](https://github.com/pytorch/pytorch/blob/75bf5f2b59148ea8b6a3844248bcc645dd80e0c8/torch/jit/annotations.py#L332))? it seems what this part of code is doing is to try to recursively compile the class that does not have class annotations?

Reply:
Yes, it is trying to recursively compile the class of a module attribute even if it is not annotated.